llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp

```diff
@@ -4310,44 +4310,349 @@ AMDGPUInstructionSelector::selectVOP3NoMods(MachineOperand &Root) const {
+    if ((HiStat == IS_SAME || HiStat == IS_NEG) &&
+        (LoStat == IS_SAME || LoStat == IS_NEG) &&
+        isInlinableConstant(*NewOp, TII)) {
+      if (HiStat == IS_NEG)
+        Mods ^= SISrcMods::NEG_HI;
+      if (LoStat == IS_NEG)
+        Mods ^= SISrcMods::NEG;
+      // opsel = opsel_hi = 0, since the upper half and lower half both
+      // the same as the target inlinable constant
+      return true;
+    }
+  }
+  return false;
+}
+
+std::pair<const MachineOperand *, unsigned>
+AMDGPUInstructionSelector::selectVOP3PModsImpl(const MachineOperand *Op,
+                                               const MachineRegisterInfo &MRI,
+                                               bool IsDOT) const {
+  unsigned Mods = 0;
+  const MachineOperand *RootOp = Op;
+  std::pair<const MachineOperand *, SrcStatus> Stat =
+      getSrcStats(Op, MRI, true)[0];
+  if (!Stat.first->isReg()) {
+    Mods |= SISrcMods::OP_SEL_1;
+    return {Op, Mods};
+  }
+  if (Stat.second == IS_NEG)
     Mods ^= (SISrcMods::NEG | SISrcMods::NEG_HI);
-    Src = MI->getOperand(1).getReg();
-    MI = MRI.getVRegDef(Src);
+
+  Op = Stat.first;
+  MachineInstr *MI = MRI.getVRegDef(Op->getReg());
+
+  if (MI->getOpcode() != AMDGPU::G_BUILD_VECTOR || MI->getNumOperands() != 3 ||
+      (IsDOT && Subtarget->hasDOTOpSelHazard())) {
+    Mods |= SISrcMods::OP_SEL_1;
+    return {Op, Mods};
   }
 
-  // TODO: Handle G_FSUB 0 as fneg
+  SmallVector<std::pair<const MachineOperand *, SrcStatus>> Statlist_Hi;
+  Statlist_Hi = getSrcStats(&MI->getOperand(2), MRI);
 
-  // TODO: Match op_sel through g_build_vector_trunc and g_shuffle_vector.
-  (void)IsDOT; // DOTs do not use OPSEL on gfx942+, check ST.hasDOTOpSelHazard()
+  if (Statlist_Hi.size() == 0) {
+    Mods |= SISrcMods::OP_SEL_1;
+    return {Op, Mods};
+  }
+
+  SmallVector<std::pair<const MachineOperand *, SrcStatus>> Statlist_Lo;
+  Statlist_Lo = getSrcStats(&MI->getOperand(1), MRI);
+
+  if (Statlist_Lo.size() == 0) {
+    Mods |= SISrcMods::OP_SEL_1;
+    return {Op, Mods};
+  }
 
+  for (int i = Statlist_Hi.size() - 1; i >= 0; i--) {
+    for (int j = Statlist_Lo.size() - 1; j >= 0; j--) {
+      if (isSameOperand(Statlist_Hi[i].first, Statlist_Lo[j].first)) {
+        if (validToPack(Statlist_Hi[i].second, Statlist_Lo[j].second, Mods,
+                        Statlist_Hi[i].first, RootOp, TII, MRI))
+          return {Statlist_Hi[i].first, Mods};
+      }
+    }
+  }
   // Packed instructions do not have abs modifiers.
   Mods |= SISrcMods::OP_SEL_1;
 
-  return std::pair(Src, Mods);
+  return {Op, Mods};
+}
+
+int64_t getAllKindImm(const MachineOperand *Op) {
+  switch (Op->getType()) {
+  case MachineOperand::MachineOperandType::MO_Immediate:
+    return Op->getImm();
+  case MachineOperand::MachineOperandType::MO_CImmediate:
+    return Op->getCImm()->getSExtValue();
+  case MachineOperand::MachineOperandType::MO_FPImmediate:
+    return Op->getFPImm()->getValueAPF().bitcastToAPInt().getSExtValue();
+  }
```

Comment:
do a default with llvm_unreachable; otherwise a warning will be emitted here IIUC.